### PR TITLE
NEW: Add DFBRegistration Function

### PR DESCRIPTION
**Pull request by ruqayya**
To-do-list:
- [x] DFBR on tissue 
- [x] Excluding matching points that lie outside tissue boundary
- [x] DFBR on prealigned tissue sections in a blockwise manner

### CODE REVIEW

File: tests/test_wsi_registration.py [hunk 1]
```diff
@@ -219,3 +219,22 @@ def test_match_histograms():
...
+    moving_img = cv2.warpAffine(
+        moving_img, pre_transform[0:-1][:], fixed_img.shape[:2][::-1]
+    )
+    moving_msk = cv2.warpAffine(
+        moving_msk, pre_transform[0:-1][:], fixed_img.shape[:2][::-1]
+    )
+
+    df = DFBRegister()
+    _ = df.register(fixed_img, moving_img, fixed_msk, moving_msk)
+    print("hello")
```
Comment (shaneahmed) [inline tests/test_wsi_registration.py:240]:
```suggestion

```

File: tests/test_wsi_registration.py [hunk 2]
```diff
@@ -219,3 +219,22 @@ def test_match_histograms():
...
+    pre_transform = np.array([[-1, 0, 337.8], [0, -1, 767.7], [0, 0, 1]])
+    moving_img = cv2.warpAffine(
+        moving_img, pre_transform[0:-1][:], fixed_img.shape[:2][::-1]
+    )
+    moving_msk = cv2.warpAffine(
+        moving_msk, pre_transform[0:-1][:], fixed_img.shape[:2][::-1]
+    )
+
+    df = DFBRegister()
+    _ = df.register(fixed_img, moving_img, fixed_msk, moving_msk)
```
Comment (shaneahmed) [inline tests/test_wsi_registration.py:239]: This needs to return an image, transformation or a metric which should be compared against GT.

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 3]
```diff
@@ -539,3 +542,314 @@ def estimate_affine_transform(
         matrix[-1, :] = [0, 0, 1]
 
         return matrix
+
+    @staticmethod
+    def get_tissue_regions(
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:547]: Can we not do it using WSIReader?

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 4]
```diff
@@ -539,3 +542,314 @@ def estimate_affine_transform(
...
+        return (
+            fixed_tissue_image,
+            fixed_tissue_mask,
+            moving_tissue_image,
+            moving_tissue_mask,
+            (minr, minc, maxr, maxc),
+        )
+
+    @staticmethod
+    def find_points_inside_boundary(mask: np.ndarray, points: np.ndarray):
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:614]: I understand similar functions have been implemented in patch extraction.

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 5]
```diff
@@ -539,3 +542,314 @@ def estimate_affine_transform(
...
+        bound_points, _ = cv2.findContours(mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
+
+        included_points_ind = []
+        for bound_p in bound_points:
+            bound_p = path.Path(np.squeeze(bound_p))
+            ind = bound_p.contains_points(points).nonzero()
+            included_points_ind = np.hstack([included_points_ind, ind[0]])
+        return included_points_ind.astype(int)
+
+    def filtering_matching_points(
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:642]: Is there not something similar available in cv2 or sklearn? This should be available in existing libraries I guess.

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 6]
```diff
@@ -539,3 +542,314 @@ def estimate_affine_transform(
...
+            fixed_matched_points, moving_matched_points
+        )
+        moving_tissue = cv2.warpAffine(
+            moving_tissue, tissue_transform[0:-1][:], fixed_tissue.shape[:2][::-1]
+        )
+        moving_mask = cv2.warpAffine(
+            moving_mask, tissue_transform[0:-1][:], fixed_tissue.shape[:2][::-1]
+        )
+
+        blocks_bbox = [
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:776]:
```suggestion
        blocks_bounding_box = [
```

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 7]
```diff
@@ -539,3 +542,314 @@ def estimate_affine_transform(
...
+            [
+                np.ceil(fixed_tissue.shape[0] / 2),
+                fixed_tissue.shape[0],
+                np.ceil(fixed_tissue.shape[1] / 2),
+                fixed_tissue.shape[1],
+            ],
+        ]
+        fixed_matched_points, moving_matched_points, quality = [], [], []
+        for _index, block in enumerate(blocks_bbox):
+            bbox = [int(x) for x in block]
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:804]:
```suggestion
            bounding_box = [int(x) for x in block]
```

Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py:804]: also shouldn't the block should already be defined as int to avoid an errors?

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 5 above)
Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py:642]: https://docs.opencv.org/4.x/dc/dc3/tutorial_py_matcher.html

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 3 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py:547]: I don't think we can use it for two reasons. First, we don't have a function in WSIReader that extracts tissue regions using a binary mask. Second, get_tissue_regions function would be called on a pre-aligned pair of images since pre-alignment is mandatory for DFBR for a successful registration. And, using WSIReader we cannot apply the geometrical transformation.

File: tests/test_wsi_registration.py (hunk 2 above)
Comment (ruqayya) [reply tests/test_wsi_registration.py]: https://github.com/TissueImageAnalytics/tiatoolbox/blob/7be8b4c7387ead7ee0cf082f14ded5ab87bc771a/tests/test_wsi_registration.py#L238-L244

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 6 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py]: https://github.com/TissueImageAnalytics/tiatoolbox/blob/7be8b4c7387ead7ee0cf082f14ded5ab87bc771a/tiatoolbox/tools/registration/wsi_registration.py#L776

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 7 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py]: https://github.com/TissueImageAnalytics/tiatoolbox/blob/7be8b4c7387ead7ee0cf082f14ded5ab87bc771a/tiatoolbox/tools/registration/wsi_registration.py#L776-L804

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 3 above)
Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py:547]: How about PatchExtraction?
https://github.com/TissueImageAnalytics/tiatoolbox/blob/62dc5bdb3e9c9359b667f48da80073285d1772f9/tiatoolbox/tools/patchextraction.py#L117

File: tests/test_wsi_registration.py [hunk 8]
```diff
@@ -219,3 +202,40 @@ def test_match_histograms():
...
+        match=r".*The required shape for fixed and moving images is n x m x 3.*",
+    ):
+        _ = df.register(fixed_img[:, :, 0], moving_img[:, :, 0], fixed_msk, moving_msk)
+
+    with pytest.raises(
+        ValueError, match=r".*The input images are expected to have 3 channels.*"
+    ):
+        _ = df.register(
+            fixed_img[:, :, :1], moving_img[:, :, :1], fixed_msk, moving_msk
+        )
```
Comment (John-P) [inline tests/test_wsi_registration.py:225]: Usually fail cases would be a seperate unit test

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 9]
```diff
@@ -6,7 +6,9 @@
 import scipy.ndimage as ndi
 import torch
 import torchvision
+from matplotlib import path
 from skimage import exposure, filters
+from skimage.measure import regionprops
```
Comment (John-P) [inline tiatoolbox/tools/registration/wsi_registration.py:11]: Consider replacing with cv2 version for speed

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 10]
```diff
@@ -536,6 +531,356 @@ def estimate_affine_transform(
...
+        )
+        tissue_transform = np.vstack([tissue_transform, np.array([0, 0, 1])])
+        block_transform = np.vstack([block_transform, np.array([0, 0, 1])])
+        return np.matmul(
+            np.matmul(
+                np.matmul(np.matmul(translation, block_transform), tissue_transform),
+                translation_,
+            ),
+            transform_initializer,
+        )
```
Comment (John-P) [inline tiatoolbox/tools/registration/wsi_registration.py:886]: You might want to break this up a bit for readability. It's a bit hard to parse like this.

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 11]
```diff
@@ -536,6 +531,356 @@ def estimate_affine_transform(
...
+                [0, 0, 1],
+            ]
+        )
+        translation = np.array(
+            [
+                [1, 0, tissue_top_left_coord[1]],
+                [0, 1, tissue_top_left_coord[0]],
+                [0, 0, 1],
+            ]
+        )
```
Comment (John-P) [inline tiatoolbox/tools/registration/wsi_registration.py:877]: Why is there a 'translation_' and a 'translation' variable here?

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 12]
```diff
@@ -536,6 +531,356 @@ def estimate_affine_transform(
...
+                fixed_matched_points, duplicate_ind, axis=0
+            )
+            moving_matched_points = np.delete(
+                moving_matched_points, duplicate_ind, axis=0
+            )
+            quality = np.delete(quality, duplicate_ind)
+
+        return fixed_matched_points, moving_matched_points, quality
+
+    def register(
```
Comment (John-P) [inline tiatoolbox/tools/registration/wsi_registration.py:712]: This function could do with some comments explaining what is going on. Also, a bit of a general tidy up. Maybe better variable names, or you may be able to extract some parts into another function to make this one less complex.

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 10 above)
Comment (John-P) [reply tiatoolbox/tools/registration/wsi_registration.py:886]: Not that you can also use `@` for matrix multiplication. e.g. `X @ t`

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 4 above)
Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py:605]: https://tia-toolbox.readthedocs.io/en/latest/_modules/tiatoolbox/wsicore/wsireader.html#VirtualWSIReader

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 5 above)
Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py:633]: Explore this option in parallel and create a PR later.

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 9 above)
Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py:11]: https://docs.opencv.org/3.4/de/d62/tutorial_bounding_rotated_ellipses.html

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 13]
```diff
@@ -536,6 +531,356 @@ def estimate_affine_transform(
...
+        self,
+        fixed_img: np.ndarray,
+        moving_img: np.ndarray,
+        fixed_mask: np.ndarray,
+        moving_mask: np.ndarray,
+        transform_initializer: np.ndarray = None,
+    ) -> np.ndarray:
+        """Image Registration.
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:720]:
```suggestion
        """Performs whole slide image registration.
```

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 14]
```diff
@@ -536,6 +531,356 @@ def estimate_affine_transform(
...
+            fixed_matched_points,
+            moving_matched_points,
+            quality,
+        )
+
+        tissue_transform = DFBRegister.estimate_affine_transform(
+            fixed_matched_points, moving_matched_points
+        )
+
+        moving_tissue = cv2.warpAffine(
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:790]:
```suggestion
        # Apply Affine transformation
        moving_tissue = cv2.warpAffine(
```

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 9 above)
Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py:11]: https://www.pythonpool.com/cv2-boundingrect/

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 4 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py:647]: Now using PatchExtractor.filter_coordinates_fast.
https://github.com/TissueImageAnalytics/tiatoolbox/blob/3225a5835708a89cd5999385b406aa3d1a663fe8/tiatoolbox/tools/registration/wsi_registration.py#L647-L673

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 9 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py]: Now using opencv.
https://github.com/TissueImageAnalytics/tiatoolbox/blob/3225a5835708a89cd5999385b406aa3d1a663fe8/tiatoolbox/tools/registration/wsi_registration.py#L617-L620

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 10 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py]: https://github.com/TissueImageAnalytics/tiatoolbox/blob/3225a5835708a89cd5999385b406aa3d1a663fe8/tiatoolbox/tools/registration/wsi_registration.py#L1038-L1040

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 11 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py]: Renamed these variables.
https://github.com/TissueImageAnalytics/tiatoolbox/blob/3225a5835708a89cd5999385b406aa3d1a663fe8/tiatoolbox/tools/registration/wsi_registration.py#L1024-L1037

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 13 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py]: https://github.com/TissueImageAnalytics/tiatoolbox/blob/3225a5835708a89cd5999385b406aa3d1a663fe8/tiatoolbox/tools/registration/wsi_registration.py#L925-L933

File: tests/test_wsi_registration.py (hunk 8 above)
Comment (ruqayya) [reply tests/test_wsi_registration.py]: https://github.com/TissueImageAnalytics/tiatoolbox/blob/e9f33b468d301ebd6c55c3f86bdc9eb03ef6dbcd/tests/test_wsi_registration.py#L241-L292

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 14 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py]: https://github.com/TissueImageAnalytics/tiatoolbox/blob/a040932dda0d4641e69f41d2b19b7f85c16e7ca7/tiatoolbox/tools/registration/wsi_registration.py#L808-L811

Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py]: Removed Affine from the comment as I was getting a spelling error in pre-commit. 
https://github.com/TissueImageAnalytics/tiatoolbox/blob/695a97a48e399b7d19d646ab8d7a6322c8356e13/tiatoolbox/tools/registration/wsi_registration.py#L808-L811

**Review by shaneahmed — APPROVED** [review]
Thanks @ruqayya . This looks great now.
